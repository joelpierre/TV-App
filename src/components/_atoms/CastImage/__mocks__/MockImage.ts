import { ITvShowImage } from '../../../../types/interfaces';

export const MockImage: ITvShowImage = {
  medium: "http://static.tvmaze.com/uploads/images/medium_portrait/130/325669.jpg",
  original: "http://static.tvmaze.com/uploads/images/original_untouched/130/325669.jpg"
};
