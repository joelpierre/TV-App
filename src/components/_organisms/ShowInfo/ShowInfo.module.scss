.ShowInfo {
  $root: &;
  width: 100%;
  position: relative;
  padding: spacing(6) 0;
  overflow: hidden;

  @include breakpoint('md') {
    padding: spacing(12) 0;
  }

  &__container {
    @include max-width-container;
    display: flex;
    flex-wrap: wrap;

    @include breakpoint('md') {
      flex-wrap: nowrap;
    }
  }

  &__heading {
    font-size: 30px;
    font-family: $font-regular;
    margin-bottom: spacing(5);

    @include breakpoint('md') {
      margin-bottom: spacing(7);
    }
  }

  &__details,
  &__cast {
    width: 100%;
    margin-bottom: spacing(5);

    @include breakpoint('md') {
      width: 50%;
      margin-bottom: 0;
    }
  }

  &__details {
    @include breakpoint('md') {
      padding-right: spacing(6);
    }
  }

  &__cast {
    margin-bottom: 0;
  }

  &__table {
    width: 100%;
    font-size: 20px;

    &-body {
      display: flex;
      flex-wrap: wrap;
    }

    &-row {
      display: flex;
      align-items: flex-start;
      justify-content: flex-start;
      padding: 0;
      width: 50%;
      flex-wrap: wrap;
      flex-direction: column;
      margin: 0 0 spacing(3);

      @include breakpoint('sm') {
        margin-bottom: 0;
        height: 100px;
      }

      @include breakpoint('md') {
        border-bottom: 1px solid $color-brand;
        flex-wrap: nowrap;
        align-items: center;
        justify-content: center;
        width: 100%;
        flex-direction: row;
      }
    }

    &-cell {
      width: 100%;
      display: flex;
      align-items: flex-start;
      flex-wrap: wrap;
      font-family: $font-light;

      @include breakpoint('md') {
        padding: 0 spacing(2);
      }
    }

    &-cell--image {
      max-width: 58px;

      @include breakpoint('md') {
        max-width: 74px;
      }

      @include breakpoint('md') {
        max-width: 90px;
        margin-right: spacing(6);
        overflow: hidden;
      }
    }

    &-heading {
      width: 100%;

      @include breakpoint('md') {
        max-width: 214px; // From Sketch File
      }
    }
  }

  &__table-body--cast {
    #{$root} {
      &__table-row {
        width: 100%;
        flex-wrap: nowrap;
        flex-direction: row;
        align-items: center;
        height: 80px;
        margin-bottom: 0;

        @include breakpoint('sm') {
          height: 100px;
        }

        @include breakpoint('md') {
          width: 100%;
          align-items: center;
          justify-content: flex-start;
        }
      }

      &__table-heading {
        width: 100%;
        margin-left: spacing(3);

        @include breakpoint('md') {
          display: flex;
          margin-left: 0;
          max-width: none;
        }
      }

      &__real-name {
        font-weight: normal;
        font-family: $font-light;
        display: block;

        @include breakpoint('md') {
          margin-left: auto;
          display: inline;
        }
      }
    }
  }
}
