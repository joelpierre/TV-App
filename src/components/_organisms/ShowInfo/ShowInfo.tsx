import React from 'react';
import classNames from 'classnames';

import styles from './ShowInfo.module.scss';
import { ITvShow, ITvShowCast, ITvShowImage } from 'common/types/interfaces';
import CastImage from '@atoms/CastImage';

export interface IShowInfoProps {
  className?: string;
  show: ITvShow;
}

interface IShowInfo {
  heading: string;
  content: string;
}

interface IShowCast {
  image?: ITvShowImage;
  characterName: string;
  realName: string;
}

const getCastInfo = (cast: ITvShowCast[]): IShowCast[] => {
  if (!cast) {
    return [];
  }
  return cast.map(people => {
    return {
      image: people.person.image,
      characterName: people.character.name,
      realName: people.person.name
    };
  });
};

const ShowInfo: React.FunctionComponent<IShowInfoProps> = ({
  className,
  show: { name, network, status, genres, _embedded }
}) => {
  const showInfo: IShowInfo[] = [
    { heading: 'Streamed On', content: network?.name },
    { heading: 'Schedule', content: name },
    { heading: 'Status', content: status },
    { heading: 'Genres', content: genres?.join(', ') }
  ];

  const castInfo = getCastInfo(_embedded?.cast);

  return (
    <div className={classNames(styles.ShowInfo, className)}>
      <div className={styles.ShowInfo__container}>
        <div className={styles.ShowInfo__details}>
          <h2 className={styles.ShowInfo__heading}>
            Show Info
          </h2>

          <table data-testid="show-info" className={styles.ShowInfo__table}>
            <tbody className={styles['ShowInfo__table-body']}>
              {showInfo.map((info, index) => {
                if (!info.content) {
                  return null;
                }
                return (
                  <tr key={`${info} ${index}`} className={styles['ShowInfo__table-row']}>
                    <th className={styles['ShowInfo__table-heading']}>{info.heading}</th>
                    <td className={styles['ShowInfo__table-cell']}>{info.content}</td>
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>

        {castInfo.length > 0 && (
          <div className={styles.ShowInfo__cast}>
            <h2 className={`${styles.ShowInfo__heading} ${styles['ShowInfo__heading--cast']}`}>
              Starring
            </h2>

            <table data-testid="show-cast" className={styles.ShowInfo__table}>
              <tbody className={`${styles['ShowInfo__table-body']} ${styles['ShowInfo__table-body--cast']}`}>
                {castInfo.map((actor, index) => (
                  <tr key={`${actor.realName} ${index}`} className={styles['ShowInfo__table-row']}>
                    <td className={`${styles['ShowInfo__table-cell']} ${styles['ShowInfo__table-cell--image']}`}>
                      <CastImage image={actor.image} name={actor.realName} />
                    </td>
                    <th className={styles['ShowInfo__table-heading']}>
                      {actor.characterName}
                      <span className={styles['ShowInfo__real-name']}>
                        {actor.realName}
                      </span>
                    </th>
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
        )}
      </div>
    </div>
  );
};

export default ShowInfo;
