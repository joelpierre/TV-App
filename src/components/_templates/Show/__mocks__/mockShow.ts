/* eslint-disable */
import { ITvShow } from '../../../../types/interfaces';
import { MockImage } from '../../../_atoms/CastImage/__mocks__/MockImage';

export const MockShow: ITvShow = {
  "id": 31534,
  "url": "http://www.tvmaze.com/shows/31534/the-russell-howard-hour",
  "name": "The Russell Howard Hour",
  "type": "Talk Show",
  "language": "English",
  "genres": [
    "Comedy"
  ],
  "status": "Running",
  "runtime": 60,
  "premiered": "2017-09-21",
  "officialSite": "https://www.sky.com/watch/title/series/89419c88-0079-426a-b29b-bc28bc403a58/the-russell-howard-hour",
  "schedule": {
    "time": "22:00",
    "days": [
      "Thursday"
    ]
  },
  "rating": {
    "average": 8.4
  },
  "weight": 83,
  "network": {
    "id": 63,
    "name": "Sky 1",
    "country": {
      "name": "United Kingdom",
      "code": "GB",
      "timezone": "Europe/London"
    }
  },
  "webChannel": null,
  "externals": {
    "tvrage": null,
    "thetvdb": 334902,
    "imdb": "tt7544016"
  },
  "image": MockImage,
  "summary": "<p><b>The Russell Howard Hour</b> is a weekly talk show in which Russell Howard will be chatting with celebrity guests alongside comedy stand-up. </p>",
  "updated": 1603404686,
  "_links": {
    "self": {
      "href": "http://api.tvmaze.com/shows/31534"
    },
    "previousepisode": {
      "href": "http://api.tvmaze.com/episodes/1955660"
    },
    "nextepisode": {
      "href": "http://api.tvmaze.com/episodes/1955661"
    }
  },
  "_embedded": {
    "cast": [
      {
        "person": {
          "id": 64866,
          "url": "http://www.tvmaze.com/people/64866/russell-howard",
          "name": "Russell Howard",
          "country": {
            "name": "United Kingdom",
            "code": "GB",
            "timezone": "Europe/London"
          },
          "birthday": "1980-03-23",
          "deathday": null,
          "gender": "Male",
          "image": {
            "medium": "http://static.tvmaze.com/uploads/images/medium_portrait/175/437607.jpg",
            "original": "http://static.tvmaze.com/uploads/images/original_untouched/175/437607.jpg"
          },
          "_links": {
            "self": {
              "href": "http://api.tvmaze.com/people/64866"
            }
          }
        },
        "character": {
          "id": 456592,
          "url": "http://www.tvmaze.com/characters/456592/the-russell-howard-hour-host",
          "name": "Host",
          "image": null,
          "_links": {
            "self": {
              "href": "http://api.tvmaze.com/characters/456592"
            }
          }
        },
        "self": true,
        "voice": false
      }
    ]
  }
};
