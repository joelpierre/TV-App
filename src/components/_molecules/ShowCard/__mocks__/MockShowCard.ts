import { ITvSchedule } from '../../../../types/interfaces';
import { MockShow } from '../../../_templates/Show/__mocks__/mockShow';
import { MockImage } from '../../../_atoms/CastImage/__mocks__/MockImage';

export const MockShowCard: ITvSchedule = {
  url: '',
  type: '',
  summary: '',
  season: 2,
  runtime: 50,
  number: 2,
  name: 'Test Show Card',
  image: MockImage.original,
  id: 123,
  airtime: '09:00',
  airstamp: 'airStamp',
  airdate: '01-01-2020',
  show: MockShow
};
