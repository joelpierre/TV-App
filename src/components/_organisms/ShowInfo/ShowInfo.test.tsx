import React from 'react';
import { render, RenderResult } from 'test-utils';

import ShowInfo, { IShowInfoProps } from './ShowInfo';
import { MockShow } from '../../_templates/Show/__mocks__/mockShow';

const defaultProps: IShowInfoProps = {
  show: MockShow
};

const setup = (props: Partial<IShowInfoProps> = {}): RenderResult => {
  const setupProps = { ...defaultProps, ...props };
  return render(<ShowInfo {...setupProps} />);
};

describe('<ShowInfo/>', () => {
  let wrapper: RenderResult;

  describe('Default State', () => {
    beforeEach(() => {
      wrapper = setup();
    });

    it('Should render without crashing', () => {
      expect(wrapper.container).not.toBeEmptyDOMElement();
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });

    it('should render two h2 tags', () => {
      expect(wrapper.container.getElementsByTagName('h2')).toHaveLength(2);
    });

    it('should render two table tags', () => {
      expect(wrapper.getByTestId('show-info')).toBeTruthy();
      expect(wrapper.getByTestId('show-cast')).toBeTruthy();
    });
  });

  describe('Given no cast info', () => {
    beforeEach(() => {
      wrapper = setup({ show: { ...MockShow, _embedded: undefined } });
    });

    it('should render one h2 tag', () => {
      expect(wrapper.container.getElementsByTagName('h2')).toHaveLength(1);
    });

    it('should not render cast table', () => {
      expect(wrapper.getByTestId('show-info')).toBeTruthy();
      expect(wrapper.queryByTestId('show-cast')).toBeFalsy();
    });
  });
});
