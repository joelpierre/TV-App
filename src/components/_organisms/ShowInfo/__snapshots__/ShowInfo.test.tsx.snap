// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShowInfo/> Default State should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ShowInfo"
  >
    <div
      class="ShowInfo__container"
    >
      <div
        class="ShowInfo__details"
      >
        <h2
          class="ShowInfo__heading"
        >
          Show Info
        </h2>
        <table
          class="ShowInfo__table"
          data-testid="show-info"
        >
          <tbody
            class="ShowInfo__table-body"
          >
            <tr
              class="ShowInfo__table-row"
            >
              <th
                class="ShowInfo__table-heading"
              >
                Streamed On
              </th>
              <td
                class="ShowInfo__table-cell"
              >
                Sky 1
              </td>
            </tr>
            <tr
              class="ShowInfo__table-row"
            >
              <th
                class="ShowInfo__table-heading"
              >
                Schedule
              </th>
              <td
                class="ShowInfo__table-cell"
              >
                The Russell Howard Hour
              </td>
            </tr>
            <tr
              class="ShowInfo__table-row"
            >
              <th
                class="ShowInfo__table-heading"
              >
                Status
              </th>
              <td
                class="ShowInfo__table-cell"
              >
                Running
              </td>
            </tr>
            <tr
              class="ShowInfo__table-row"
            >
              <th
                class="ShowInfo__table-heading"
              >
                Genres
              </th>
              <td
                class="ShowInfo__table-cell"
              >
                Comedy
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="ShowInfo__cast"
      >
        <h2
          class="ShowInfo__heading ShowInfo__heading--cast"
        >
          Starring
        </h2>
        <table
          class="ShowInfo__table"
          data-testid="show-cast"
        >
          <tbody
            class="ShowInfo__table-body ShowInfo__table-body--cast"
          >
            <tr
              class="ShowInfo__table-row"
            >
              <td
                class="ShowInfo__table-cell ShowInfo__table-cell--image"
              >
                <div
                  class="CastImage"
                >
                  <img
                    alt="Russell Howard image"
                    class="CastImage__image"
                    loading="lazy"
                    src="//static.tvmaze.com/uploads/images/medium_portrait/175/437607.jpg"
                    width="58"
                  />
                </div>
              </td>
              <th
                class="ShowInfo__table-heading"
              >
                Host
                <span
                  class="ShowInfo__real-name"
                >
                  Russell Howard
                </span>
              </th>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
